### DISCUSSION TITLE
Date Picker

### DISCUSSION THREAD

**Discussion by Rozerinay** [nl-design-system/backlog#272] [Component suggestions]
<img width="720" alt="visuele weergave van het date picker component" src="https://github.com/nl-design-system/backlog/assets/84851545/c9072f13-9482-4258-8e1b-7819e26fd77f">

We willen graag meer leren over het Date Picker component. In welke vorm, of vormen, wordt deze momenteel ingezet en toegepast binnen jouw organisatie?

Deel hieronder een link naar Figma, Storybook, Github of screenshots. Met deze informatie kunnen we bewijzen dat het component onderdeel moet worden van de NL Design System component library.

Daarnaast wordt het delen van onderzoek gewaardeerd. Dus heb je zelf onderzoek gedaan en zijn er inzichten voor dit component? Dan horen we dit ook graag!

[Link naar bericht in Slack](https://codefornl.slack.com/archives/C7KDQRXMZ/p1718791910428269)

---

## Naam
Date Picker

## Doel
Invoerveld om een beschikbare datum te kiezen via een op maat gemaakte kalender.

## Gerelateerde componenten
[Calendar](https://github.com/orgs/nl-design-system/discussions/273), [Date Input](https://github.com/orgs/nl-design-system/discussions/270), [Date Input Group](https://github.com/orgs/nl-design-system/discussions/271)

## Onderzoek

[Accessible datepickers - Axesslab](https://axesslab.com/accessible-datepickers/)
Artikel waarin enkele criteria voor een toegankelijke Date Picker worden benoemd. Aangevuld met een lijst naar enkele voorbeelden.

[Cally - Nick Williams](https://wicky.nillia.ms/cally/)
Voorbeeld van een Date Picker component waar [toegankelijkheid](https://wicky.nillia.ms/cally/accessibility/) specifiek wordt benoemd.

[Designing The Perfect Date And Time Picker - Smashing magazine](https://www.smashingmagazine.com/2017/07/designing-perfect-date-time-picker/)
Uitgebreid artikel waarin de zoektocht naar de perfecte Data (and Time) Picker wordt beschreven.

[What makes an accessible date picker? Is it even possible? - Russ Weakley - Youtube](https://m.youtube.com/watch?v=D2Gy2WN4Iys)
Video waarin wordt uitgelegd wat een Date Picker toegankelijk maakt en inzoomt op 3 verschillende Date Pickers, waaronder de native Date Picker. 

[37 ‘Date Picker’ Design Examples](https://baymard.com/ecommerce-design-examples/date-picker)
Onderzoek toont aan dat wanneer de beschikbaarheid niet adequaat of accuraat wordt gecommuniceerd in de kalender, gebruikers extra tijd en moeite moeten besteden aan het verifiëren van deze informatie, wat leidt tot frustratie en teleurstelling – en mogelijke verlating.

[Datepicker Input - NS Design System](https://design.ns.nl/4a05a30ad/p/30e5a5-datepicker-input)
Implementatie van de Datepicker Input component bij het NS Design System.

[Date Picker - Ministry of Justice Design System](https://design-patterns.service.justice.gov.uk/components/date-picker/)
Implementatie van de Datepicker Input component bij Ministry of Justice. Inclusief documentatie over gebruik, toegankelijkheid en [GitHub Discussion](https://github.com/ministryofjustice/moj-frontend/discussions/703)

**Comment by MariekeBrouwer** [comment]
Dit is de standaard optie voor Open Formulieren die Groningen gebruikt bij een datum/ datum/tijd selectie. Dus zowel voor selectie van data voor evenementen als bijvoorbeeld geboortedatum wordt dan deze optie gebruikt. Bij andere systemen (zoals qmatic) gebruiken we voor de geboortedatum ook wel het date input group component. Voor ons is de keuze nu eigenlijk meer systeem afhankelijk... 

![image](https://github.com/nl-design-system/backlog/assets/149144374/84a8e045-dc9f-45a4-b93f-84cd37550fe2)

![image](https://github.com/nl-design-system/backlog/assets/149144374/fbcf3f13-af0e-475e-84e2-6ddf449a6161)

**Comment by jeffreylauwers** [comment]
## Den Haag

- [Storybook](https://nl-design-system.github.io/denhaag/?path=/story/open-forms-date-field--default)
- [GitHub](https://github.com/nl-design-system/denhaag/blob/main/packages/storybook/src/open-forms/DateField.stories.tsx)
- [Figma (Huidig)](https://www.figma.com/design/JpoY3waVoQGlLQzQXTL9nn/%F0%9F%8C%BB-Den-Haag-Design-System?node-id=1681-7756&t=mYo78jne5kP62APW-4)

---

**Storybook**

Default - geen datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 10 59 36](https://github.com/user-attachments/assets/bb0c843a-5a26-4abd-8dc0-742aa48d12b3)

Active - geen datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 10 59 43](https://github.com/user-attachments/assets/ea05950e-1831-4e49-9db6-d99dc6a335bb)

Active - hover state op een dag
![Scherm­afbeelding 2024-10-18 om 11 07 08](https://github.com/user-attachments/assets/bc398701-a866-426c-b0b9-8cdc152df854)

Default - datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 54 30](https://github.com/user-attachments/assets/3507fdb2-f822-42b7-9cd9-4c179b4e49d6)

Active - datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 54 57](https://github.com/user-attachments/assets/9e79ef84-9fa1-4b8c-8c37-780dba260d86)

Maand aanpassen
![Scherm­afbeelding 2024-10-18 om 11 07 27](https://github.com/user-attachments/assets/5cafec1b-0716-46a2-9b83-d8882626e9e7)

Jaar aanpassen
![Scherm­afbeelding 2024-10-18 om 11 07 37](https://github.com/user-attachments/assets/8c1ddb3e-7e51-4a89-823e-e6493a3f1788)

Zelf datum typen
![Scherm­afbeelding 2024-10-18 om 11 47 14](https://github.com/user-attachments/assets/5265fb5a-7b45-4f6f-8637-faad8886158c)


**Figma**

Default
![Scherm­afbeelding 2024-10-18 om 11 59 31](https://github.com/user-attachments/assets/a53baf83-c15e-42f3-8e24-9c35981641b5)

Active
![Scherm­afbeelding 2024-10-18 om 11 59 57](https://github.com/user-attachments/assets/1bb8044d-8869-4648-848e-b0fd7cdc27b8)

States voor dag
![Scherm­afbeelding 2024-10-18 om 11 59 40](https://github.com/user-attachments/assets/d1a3a607-7922-4c31-a4b4-58913681d5e5)

---

### Notities

**Storybook**
- Implementatie staat in Storybook onder de categorie 'Open Formulieren'.
- Kalender binnen de pop-over lijkt anders dan die van Open Formulieren zelf. Let daarbij op navigeren per maand of jaar.
- Placeholder format is 'dd-mm-yyyy'.
- Wanneer je zelf typt wordt er een 'input mask' toegepast om het format te ondersteunen.
- Dag element als cirkel vorm.

**Figma**
Ontwerp loopt uiteen ten opzichte van implementatie. Let daarbij op:
- Lettertype
- Kleur
- Icoon (kalender, rechts)
- Navigatie mogelijkheden
- Vorm van dag element

**Comment by jeffreylauwers** [comment]
## Open Formulieren

[Storybook](https://open-formulieren.github.io/open-forms-sdk/?path=/docs/pure-react-components-forms-datefield-datepicker--docs)

---

### Screenshots / Varianten

**Storybook**

Datepicker

Default - geen datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 12 06](https://github.com/user-attachments/assets/2094bf37-29f8-4ab9-b74a-893a032103dc)

Active - geen datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 12 18](https://github.com/user-attachments/assets/57f57c5a-3ca3-43ac-b88c-b6596994dca4)

Active - hover state op een dag
<img width="539" alt="Scherm­afbeelding 2024-10-18 om 11 49 46" src="https://github.com/user-attachments/assets/3a10011b-f805-46cf-971d-0518b8a5a6f5">

Default - datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 12 33](https://github.com/user-attachments/assets/f3c3713f-0276-4d74-8818-673b8b5e1cd4)

Active - datum geselecteerd
![Scherm­afbeelding 2024-10-18 om 11 12 44](https://github.com/user-attachments/assets/80633740-0dad-4ab2-ba25-4da7270b21f1)

Eerste focus element in de Calender
![Scherm­afbeelding 2024-10-18 om 11 12 57](https://github.com/user-attachments/assets/fd64184f-3fdb-47b7-ba39-259611f79cac)

Focus op een dag in de Calender
![Scherm­afbeelding 2024-10-18 om 11 13 10](https://github.com/user-attachments/assets/17c17e27-6c65-40bf-a90f-5d08c7fec414)

Limited range Datepicker - Active
![Scherm­afbeelding 2024-10-18 om 11 13 50](https://github.com/user-attachments/assets/2576c4ef-c3d0-4252-81e0-fd50ec622346)

Datepicker with disabled date - Active
![Scherm­afbeelding 2024-10-18 om 11 15 02](https://github.com/user-attachments/assets/506bb5b3-cc0b-45b0-8c7f-ad4761cbe6b1)

---

### Notities

- Datepicker als 1 woord
- Documentatie beschikbaar (Storybook). Inclusief 'Know issues'.
- Kalender icoon rechts.
- Placeholder geeft format d-m-jjjj aan.
- Mogelijkheid om zelf een datum te typen
- Wanneer je datum 09/12/1984 typt, wordt deze automatisch aangepast naar 9-12-1984.
- Varianten beschikbaar voor: 
  - Alle data beschikbaar
  - Range aan data beschikbaar
  - Bepaalde data niet beschikbaar

**Comment by jeffreylauwers** [comment]
## Rotterdam

- [Storybook](https://nl-design-system.github.io/rotterdam/?path=/docs/rods-date-picker--docs)
- [GitHub](https://github.com/nl-design-system/rotterdam/tree/main/packages/web-components/date-picker)
- [Figma](https://www.figma.com/design/ZWSC4gCrOXRUR9UX3aoZ8x/Gemeente-Rotterdam-Design-System?node-id=521-28197&t=YBcq3pZvpnJOckOZ-4)

---

### Screenshots / Varianten

**Figma**

Single 
![Scherm­afbeelding 2024-10-18 om 13 19 41](https://github.com/user-attachments/assets/1ff2bbc6-8c6f-4758-b31f-b3fcf13599f1)

Range
![Scherm­afbeelding 2024-10-18 om 13 19 46](https://github.com/user-attachments/assets/5dfef135-6518-4fe3-9703-62cb1396fbc8)

**Storybook**
<img width="982" height="643" alt="Scherm­afbeelding 2026-01-08 om 14 15 36" src="https://github.com/user-attachments/assets/66085a9c-a71a-49f1-8b65-26fce1854d32" />

---

### Notities

- Kalender icoon links
- Button rechts met label 'Kiezen'.
- Navigeren op jaar links in pop-over.
- Navigeren op maand rechts in pop-over.
- Week start bij zondag.
- Dagen worden aangegeven met 1 letter.

**Comment by jeffreylauwers** [comment]
## Utrecht

- [Storybook](https://nl-design-system.github.io/utrecht/storybook/?path=/docs/react_open-forms-date-field--docs)
- [GitHub](https://github.com/nl-design-system/utrecht/blob/main/packages/storybook-react/src/stories/open-forms/DateField.stories.tsx)

---

### Screenshots / Varianten

**Storybook**

Default
![Scherm­afbeelding 2024-10-18 om 13 32 34](https://github.com/user-attachments/assets/d997b6cd-7b37-4263-806f-1e2cf0366981)

Active
![Scherm­afbeelding 2024-10-18 om 13 32 42](https://github.com/user-attachments/assets/db47a9a1-7c01-426f-bdd8-abda1a368d43)

---

### Notities

- Implementatie staat in Storybook onder de categorie 'Open Formulieren'.
- Kalender binnen de pop-over lijkt anders dan die van Open Formulieren zelf. Let daarbij op navigeren per maand of jaar.
- Placeholder format is 'dd-mm-yyyy'.
- Wanneer je zelf typt wordt er een 'input mask' toegepast om het format te ondersteunen.

**Comment by Rozerinay** [comment]
Rian's aanvulling in het [draadje van het hiep hoi bericht](https://codefornl.slack.com/archives/C7KDQRXMZ/p1730730157181569?thread_ts=1730203204.772359&cid=C7KDQRXMZ).

> Ben geen fan. Ik begrijp dat er vraag naar is en denk dat het daarom door moet gaan, maar het moet wel retegoed getest worden op gebruikersvriendelijkheid en toegankelijkheid

**Reply by MarjonBakker** [reply]
Helemaal eens met Rian. Ik ben eigenlijk nog nooit een date picker tegengekomen waar geen toegankelijkheidsfouten in zaten. Supermooi als dat wel zou lukken natuurlijk, maar dat vereist veel werk.

**Comment by Rozerinay** [comment]
## ✨ Dit component is nu Help Wanted ✨

Help je mee hem door de Community stappen te krijgen?
[Date Picker op het Community bord](https://github.com/orgs/nl-design-system/projects/29/views/1?filterQuery=date+picker)

**Comment by jeffreylauwers** [comment]
Ideetje om als afweging toe te voegen bij dit component.

Wanneer men uit een beperkte set aan data kan kiezen (+/- 10). En deze data opties ook nog eens niet te ver in de toekomst liggen. Dan zou een [Radio Group](https://github.com/orgs/nl-design-system/discussions/222) misschien ook wel uitkomst bieden. Daarbij zou in het label dan wel iets van 'donderdag 5 december 2024' moeten staan. Waardoor de dag van de week en het jaartal duidelijk zijn.

Bonus... als men uit meerdere data kan kiezen (denk Datumprikker) dan zou een [Checkbox Group](https://github.com/orgs/nl-design-system/discussions/223) ook nog kunnen!

**Comment by VincentSmedinga** [comment]
## Amsterdam

* [Storybook](https://designsystem.amsterdam/?path=/docs/components-forms-date-picker--docs)
* [GitHub (CSS)](https://github.com/Amsterdam/design-system/blob/main/packages/css/src/components/date-picker)
* [GitHub (React)](https://github.com/Amsterdam/design-system/tree/main/packages/react/src/DatePicker)

---

### Screenshots / Varianten

#### Storybook

Standaard:
<img width="704" height="600" alt="ams-date-picker-1-default" src="https://github.com/user-attachments/assets/02a8c5f4-24bd-4456-b54a-e0c47f9cad01" />

Start- en einddatum:
<img width="704" height="600" alt="ams-date-picker-2-range" src="https://github.com/user-attachments/assets/7c366c22-af9d-4880-98d7-4c796549958b" />

Niet-selecteerbare datums:
<img width="704" height="600" alt="ams-date-picker-3-disabled" src="https://github.com/user-attachments/assets/73225afc-17c1-4fcb-9b23-eedef8102e5c" />

Minimale en maximale datum:
<img width="704" height="600" alt="ams-date-picker-4-bounds" src="https://github.com/user-attachments/assets/6214fe0f-63b5-452c-b962-d1d80d0fee45" />

In het Arabisch:
<img width="704" height="600" alt="ams-date-picker-5-arabic" src="https://github.com/user-attachments/assets/16494e94-8b6b-49fd-98d3-806e041c3f4b" />

---

## Notities

- **Maand- en jaarnavigatie** – knoppen met enkele en dubbele chevrons voor vorige/volgende maand/jaar. Custom iconen mogelijk. Datums buiten de maand worden niet getoond.
- **Datums als buttons** – dit is een formuliercomponent. Klik op een datum om deze te selecteren. In `range` modus, kies een latere datum om een bereik aan te geven. Bij klikken op een eerdere datum wordt dat de nieuwe startdatum.
- **Inline weergave** – bewust een inline component, geen dropdown of popover; een datum kiezen is een gerichte taak die ruimte en aandacht verdient. Zet hem desnoods in een Modal Dialog.
- **Disabled datums** – het is mogelijk om naar wens datums (via een functie) niet selecteerbaar te maken.
- **Huidige dag gemarkeerd** – vandaag krijgt extra nadruk en `aria-current="date"`.
- **Toegankelijk** – rendert als een `nav`-landmark met een instelbare toegankelijke naam. Schermlezers noemen voor elke dag de volledige datum inclusief weekdag. Tabnavigatie: met buttons (alle achtereenvolgens) en pijltjes (horizontaal en verticaal) volgens [WAI-ARIA grid pattern](https://www.w3.org/WAI/ARIA/apg/patterns/grid/). Een bereik wordt als zodanig voorgelezen.
- **Lokalisatie** – de `locale` bepaalt de weekdag- en maandnamen (standaard `nl-NL`). Alle talen mogelijk; de week begint op maandag. De navigatieknoppen hebben instelbare labels. Werkt met talen die van rechts naar links schrijven. Custom iconen voor buttons hebben een HTML-attribuut nodig zodat ze spiegelen.
- **‘Controlled’ component** – houdt zelf geen selectie bij: geef `value` mee en werk die bij in `onChange`.

Visueel precies hetzelfde, maar met links om te navigeren: [Calendar](https://github.com/orgs/nl-design-system/discussions/273#discussioncomment-17564959).
